### PR TITLE
Allow users to select targets via inputlist

### LINKED ISSUE #37: Ability to specify targets
A user should be able to specify targets of the given stack.yaml that should be loaded.

The emacs version calls `stack ide targets` and uses that as the list of available targets.

It would be cool to present a selection buffer of possible targets and allow the user to choose which ones they wanted. I'll have to figure out how that works.

### PR DESCRIPTION
**Pull request by parsonsmatt**
This PR modifies the `InteroSetTargets` behavior when not passed an argument to prompt the user using an `inputlist` loop.

The user is initially prompted with a list of all targets, and by either clicking or entering a number, this target's status is toggled.

This is great for toggling a small number of targets, but is extremely tedious for a large number of targets. Having related commands `InteroClearTargets` and `InteroAllTargets` would be great. Alternatively, `InteroSetTargets` could interpret the argument as a regex, and would toggle all targets matched by the regex.

Fixes #37

### CODE REVIEW

**Comment by parsonsmatt** [comment]
Alright, this is working pretty well! I like having the menu to select, and I like having the ability to pass either the target name directly or a regex that matches target names.

Apparently the newest version of `vint` breaks on anonymous functions: see [this issue](https://github.com/Kuniwak/vint/issues/227). To appease the build lords, I'm just going to use the string-functions for now as they end up being cleaner.

**Comment by parsonsmatt** [comment]
@jez you mentioned needing test projects for this -- pretty much anything that has an executable, library, and test suite should work. https://github.com/parsonsmatt/servant-persistent this one should work out.

**Comment by jez** [comment]
> this one should work out

Thanks, I'm building it now to play around.

File: autoload/intero/targets.vim [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+    " we are in a stack project, and there are desired targets. validate that
+    " they are contained inside the stack load targets
+    for l:target in a:targets
+        if index(l:stack_targets, l:target) == -1
+            " Interpret this as a regex and add all matching targets to the
+            " list.
+            let l:matches = filter(copy(l:stack_targets), 'v:val =~ l:target')
+            for l:match in l:matches
+                let l:valid_target_dict[l:match] = 1
+            endfor
```
Comment (jez) [inline autoload/intero/targets.vim]: It's cool that this uses a regex! I was thinking it'd be pretty convenient if you could just type `lib` and then there it was 👍

File: doc/intero.txt [hunk 2]
```diff
@@ -234,8 +234,14 @@ guess for which to load. You can switch between targets on the fly
...
-                  be more than one). Overwrites whatever was previously in
-                  |g:intero_load_targets|.
+                  be more than one). If the target doesn't exist in the output
+                  of stack ide targets, then it is interpreted as a regex and
+                  all matches are enabled. Overwrites whatever was previously
+                  in |g:intero_load_targets|.
+
+                  Calling |:InteroSetTargets| without any arguments will
+                  prompt you to select the targets you want to load from a
+                  list.
```
Comment (parsonsmatt) [inline doc/intero.txt]: Good call! As it happens, clear targets and use all targets do the exact same thing, so I removed clear targets.

File: autoload/intero/process.vim [hunk 3]
```diff
@@ -163,14 +164,103 @@ function! intero#process#restart() abort
...
+    " Render the target list
+    let l:menu = [l:prompt] + s:render_target_list(l:target_list)
+
+    let l:selected = 1
+    
+    let l:selected = inputlist(l:menu)
+
+    " l:selected of 0 means that the user didn't select anything, so we
+    " are done here and can return.
+    while l:selected != 0
```
Comment (parsonsmatt) [inline autoload/intero/process.vim]: Great point! Changed the logic to take that into account.

File: autoload/intero/process.vim [hunk 4]
```diff
@@ -163,14 +164,103 @@ function! intero#process#restart() abort
...
     call intero#process#restart()
 endfunction
 
 """"""""""
 " Private:
 """"""""""
 
+function! s:prompt_for_targets() abort
+    return s:inputlist_loop()
+endfunction
```
Comment (parsonsmatt) [inline autoload/intero/process.vim]: Good call, simplified.

File: autoload/intero/process.vim [hunk 5]
```diff
@@ -163,14 +164,103 @@ function! intero#process#restart() abort
 endfunction
 
 function! intero#process#restart_with_targets(...) abort
-    call intero#util#set_load_targets(a:000)
+    if a:0 == 0
+        let l:targets = s:prompt_for_targets()
+    else 
+        let l:targets = a:000
+    endif
+    call intero#targets#set_load_targets(l:targets)
```
Comment (parsonsmatt) [inline autoload/intero/process.vim]: ugh, normally I have an editorconfig to prevent that. I've included one as part of the PR.

**Pull request merged**